### DISCUSSION TITLE
comparison between zenoh vs FastDDS Discovery Server v2

### DISCUSSION THREAD

**Discussion by ros2torial** [eclipse-zenoh/roadmap#79] [Q&A]
In this article (https://zenoh.io/blog/2021-03-23-discovery/) it seems like zenoh is compared with normal DDS discovery which is definitely slow and not scalable. Is there any comparison of Zenoh with FastDDS Discovery Server v2 Discovery Data to get the latest comparison.

**Comment by ros2torial** [comment]
I experimented on comparison between zenoh vs FastDDS Discovery Server v2. In my setup there are 10 robot PC and 1 Fleet management system PC (FMS PC). In case of FastDDS Discovery Server v2, each robot PC only shares meta data like encalve info of around 424 bytes to FMS PC and gets meta data like guid info of around 488 bytes from FMS PC. This is happening every 3 seconds. So there are 2 meta messages from each robot PC to FMS PC and in total **20 meta messages** from all robot PC to FMS PC. 

In case of zenoh, each robot PC is running zenoh client and FMS PC is running zenoh router. Each zenoh client receives meta info(127 bytes) of router as well as all the other clients. so here each zenoh client recieves 10 meta messages from router and shares one meta message to router every 2.25 seconds. In total there are **11*10 meta messages**. Means it is order of n square which increases traffic. 

From this results, it seems like FastDDS Discovery Server v2 is efficient than zenoh.

**Reply by JEnoch** [reply]
If you see direct messages exchange between each Zenoh client, it means they are configured in `peer` mode (that's the default mode), where each peer tries to establish direct connectivity with each discovered Zenoh peer and Zenoh router. Those can be discovered either via scouting (by default over UDP multicast), either via gossip discovery (a router or a peer forwarding the locators from another router/peer).
Depending the behaviour of your Zenoh application, the `peer` mode could indeed cause a n-square order of traffic.
In your case, you should configure the _"zenoh client"_ on each robot in `peer` mode as explained in #82.

I guess that in your test each _"zenoh client"_ is actually a _"zenoh-bridge-dds"_, right ?
If yes, just start each bridge with `-m client` argument, or with a JSON5 config file with `mode: "client"` (see https://github.com/eclipse-zenoh/zenoh-plugin-dds/blob/master/DEFAULT_CONFIG.json5#L127)

Wrt. your measurements, the 2.25sec periodic messages you see for Zenoh are not discovery messages, but _KeepAlive_ messages that are published and subscribed by each bridge (to supervise each other and to query historical data for TRANSIENT topics when a bridge joins the group).
The period of publication can be increased via the option `--group-lease=<float>`. The default value is `3.0` seconds, implying a period of _KeepAlive_ messages of 3.0*0.75 = 2.25 seconds.
The size of _KeepAlive_ messages can be reduce setting the `--group-member-id=<string>` option to a shorter string that the default one which is an UUID in hexadecimal.

The discovery messages are usually not periodic, but are rather sent at startup and/or at of an application.
Did you also manage to measure those ones ?
For FastDDS Discovery Service I think the 3 seconds periodic messages you're measuring are SPDP messages (`Data(p)` in Wireshark) that are for discovery of participants. They are not including the biggest discovery traffic which are SEDP messages (`Data(t)`, `Data(r)` and `Data(w)` in Wireshark) that are sent at start-up for discovery of DDS Topics, Readers and Writers.
For Zenoh you should use our [Wireshark dissector](https://zenoh.io/blog/2023-01-17-zenoh-wireshark/) and look for `declare` submessages.

**Reply by ros2torial** [reply]
Hi @JEnoch,
Thanks for the information.

These are the commands I was running for the experiment -

**PC(192.168.7.17)**
`zenohd -l udp/192.168.7.17:7446`

**PC(192.168.7.250)**
`zenoh-bridge-dds -m client -e udp/192.168.7.17:7446`
`zenoh-bridge-dds -m client -e udp/192.168.7.17:7446`
`zenoh-bridge-dds -m client -e udp/192.168.7.17:7446`
`zenoh-bridge-dds -m client -e udp/192.168.7.17:7446`

Here, I was already running each zenoh-bridge-dds in client mode in single PC(192.168.7.250) and zenoh router in another PC(192.168.7.17) and these are the wireshark results -

![zenoh_discussion](https://user-images.githubusercontent.com/65568235/223206587-5aff8696-8b3f-4026-be04-fbac9cae152a.png)

each zenoh client (192.168.7.250) sends a KeepAlive message to zenoh router(192.168.7.17) and receives three messages from the zenoh router which I guess is KeepAlive message of other 3 zenoh clients. _**Why are we getting three messages from zenoh router??**_ 

**I guess that in your test each "zenoh client" is actually a "zenoh-bridge-dds", right ?**
_yes_

**The discovery messages are usually not periodic, but are rather sent at startup and/or at of an application.
Did you also manage to measure those ones ?**
_I didn't measure those messages as they were not periodic so I guess won't be an issue ._

**Reply by ros2torial** [reply]
Hi @JEnoch,
I observed that if **DDS plugin** is not used in robot side(i.e. running **zenohd** in robot), then the zenoh client on robot only receives one message from zenoh router every _2.25 sec_. But if we use DDS then the zenoh client on robot receives more message from zenoh router every _2.25 sec_. I tried many options but couldn't solve it.

**Reply by JEnoch** [reply]
> _each zenoh client (192.168.7.250) sends a KeepAlive message to zenoh router(192.168.7.17) and receives three messages from the zenoh router which I guess is KeepAlive message of other 3 zenoh clients. Why are we getting three messages from zenoh router?_

Your guess is correct: each bridge publishes a KeepAlive message which is subscribed by the 3 other bridges.
So each KeepAlive message goes to the router which routes it to the 3 other bridges.

That indeed causes a n-square traffic between the robots for mutual supervision. But that's required for the implementation of some DDS QoS such as Liveliness and Transient.
Such p2p traffic also occurs with plain DDS, but probably also using the FastDDS Discovery Service. I'm not sure why you don't see such robot-to-robot traffic using the Discovery Service. We would need to reproduce your setup to analyse.

Note we're working on replacing this KeepAlive protocol with a new Zenoh built-in [liveliness](https://github.com/eclipse-zenoh/roadmap/discussions/72) mechanism that will reduce the traffic: the mutual supervision will rely on the user data traffic and thus will not cause additional traffic.

> The discovery messages are usually not periodic, but are rather sent at startup and/or at of an application.
Did you also manage to measure those ones ?
_I didn't measure those messages as they were not periodic so I guess won't be an issue ._

Those initial discovery messages are actually the major issue that ROS2 users encounter over wireless networks or over the internet.
This initial traffic can be huge: each ROS2 node publishes discovery information for each DDS Publisher and Subscriber it has (plus 2 for each Service and 8 for each Action), and an update on `ros_discovery_info` topic for each. All those discovery messages are received by each ROS2 node.
This traffic floods the network at startup, introducing latencies and possible messages loss in wireless networks, leading to "false death" detections, and re-publications, and to more flooding...

**Reply by JEnoch** [reply]
> I observed that if DDS plugin is not used in robot side(i.e. running zenohd in robot), then the zenoh client on robot only receives one message from zenoh router every 2.25 sec. But if we use DDS then the zenoh client on robot receives more message from zenoh router every 2.25 sec. I tried many options but couldn't solve it.

Sorry, I don't understand: you confirmed that what you call "zenoh client" is actually a "zenoh-bridge-dds", right ?
So if in the robot you only run `zenohd` and not `zenoh-bridge-dds`, what's the "zenoh client" ?

Do you mind joining our Discord ? https://discord.gg/vSDSpqnbkm
So we could have a direct chat and I can better understand your experiments and results. I'm usually online on CET time zone working days.

**Reply by ros2torial** [reply]
> Those initial discovery messages are actually the major issue that ROS2 users encounter over wireless networks or over the internet.

I didn't compare the initial discovery messages between zenoh and FastDDS Discovery Server v2. Surely I will conduct this experiment too. Although in terms of periodic data case FastDDS Discovery Server v2 seems to be efficient than zenoh since it doesn't cause n-square traffic. 

I conducted an experiment last week where the periodic data of Zenoh and FastDDS Discovery Server v2 is compared -
Data exchange per second using wifi network | FastDDS Discovery Server v2 | Zenoh (peer-clients-udp)
-- | -- | -- 
One publisher(1 robot) one subscriber(FMS PC) | 516 bytes/sec | 221.888 bytes/sec
Two publisher(2 robot) one subscriber(FMS PC) | 1000 bytes/sec | 556.666666667 bytes/sec
Ten publisher(10 robot) one subscriber(FMS PC) | 5000 bytes/sec | 7298.88889 bytes/sec

As we can observe from the data that for single robot case zenoh is good although as we increase the number of robots zenoh becomes inefficient because of n-square traffic. 


> Sorry, I don't understand: you confirmed that what you call "zenoh client" is actually a "zenoh-bridge-dds", right ?
> So if in the robot you only run zenohd and not zenoh-bridge-dds, what's the "zenoh client" ?

Sorry, for confusion. Here each robot is running zenoh client by using either zenohd or zenoh-bridge-dds whichever provides better functionality and features. I am experimenting with both for now.

Also, I am unable to join discord from the above link. Will try again later.

**Reply by JEnoch** [reply]
Which applications are you running as publishers and subscriber ?
RO2 nodes? Plain DDS pub/sub? Plain Zenoh pub/sub in your experiment with zenohd?

**Reply by ros2torial** [reply]
I am ruining ROS2 nodes as publishers and subscriber. In case of single publisher and single subscriber these are the commands used for above experiment -

**FastDDS**

PC 1 (192.168.1.153)

Terminal 1
`fastdds discovery -i 0`
or
`fastdds discovery -i 0 -l 192.168.1.153 -p 11811`

Terminal 2
```
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
export ROS_DISCOVERY_SERVER="192.168.1.153:11811"
ros2 run demo_nodes_cpp listener
```

PC 2 (192.168.1.113)

Terminal 1
```
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
export ROS_DISCOVERY_SERVER="192.168.1.153:11811"
ros2 run demo_nodes_cpp talker
```


**Zenoh**

PC 1 (192.168.1.153)

Terminal 1
```
export ROS_LOCALHOST_ONLY=1
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
ROS_DOMAIN_ID=1 ros2 run demo_nodes_cpp listener
```

Terminal 2
zenoh-bridge-dds -d 1 -m peer -a /chatter -l udp/192.168.1.153:7447 --dds-localhost-only

PC 2 (192.168.1.113)

Terminal 1
```
export ROS_LOCALHOST_ONLY=1
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
ROS_DOMAIN_ID=2 ros2 run demo_nodes_cpp talker
```

Terminal 2
`zenoh-bridge-dds -d 2 -m client -a /chatter -e udp/192.168.1.153:7447 –dds-localhost-only`
